codex/blockexchange/engine/engine.nim

```diff
@@ -165,42 +192,26 @@ proc requestBlock*(
       # drop unresponsive peer
       await b.network.switch.disconnect(blockPeer.id)
 
-  trace "Sending block request to peer", peer = blockPeer.id, cid
-
   # monitor block handle
   asyncSpawn blockHandleMonitor()
 
-  # request block
-  await b.network.request.sendWantList(
-    blockPeer.id,
-    @[cid],
-    wantType = WantType.WantBlock) # we want this remote to send us a block
-
-  if (peers.len - 1) == 0:
-    trace "No peers to send want list to", cid
-    b.discovery.queueFindBlocksReq(@[cid])
-    return await blk # no peers to send wants to
-
-  # filter out the peer we've already requested from
-  let remaining = peers[1..min(peers.high, b.peersPerRequest)]
-  trace "Sending want list to remaining peers", count = remaining.len
-  for p in remaining:
-    if cid notin p.peerHave:
-      # just send wants
-      await b.network.request.sendWantList(
-        p.id,
-        @[cid],
-        wantType = WantType.WantHave) # we only want to know if the peer has the block
+  await b.sendWantHave(cid, blockPeer, peers)
```

Comment:
Changing the order here makes retrieval less efficient. `WantBlock` initiates block retrieval, while `WantHave` only signals the remotes that we want this block and doesn't trigger transmittion, we're waiting for the `WantHave` message to be sent to all peers before starting download.